src/aiida/engine/processes/control.py

```diff
@@ -241,77 +234,53 @@ def _perform_actions(
         else:
             futures[future] = process
 
-    _resolve_futures(futures, infinitive, present, wait, timeout)
+    _resolve_futures(futures, infinitive, present, timeout)
 
 
 def _resolve_futures(
     futures: dict[concurrent.futures.Future, ProcessNode],
     infinitive: str,
     present: str,
-    wait: bool = False,
     timeout: t.Optional[float] = None,
 ) -> None:
     """Process a mapping of futures representing an action on an active process.
 
     This function will echo the correct information strings based on the outcomes of the futures and the given verb
     conjugations. You can optionally wait for any pending actions to be completed before the functions returns and use a
-    timeout to put a maximum wait time on the actions.
+    timeout to put a maximum wait time on the actions. TODO fix docstring
 
     :param futures: The map of action futures and the corresponding processes.
     :param infinitive: The infinitive form of the action verb.
     :param present: The present tense form of the action verb.
-    :param wait: Set to ``True`` to wait for process response, for ``False`` the action is fire-and-forget.
     :param timeout: Raise a ``ProcessTimeoutException`` if the process does not respond within this amount of seconds.
     """
-    scheduled = {}
-
-    def handle_result(result):
-        if result is True:
-            LOGGER.report(f'request to {infinitive} Process<{process.pk}> sent')
-        elif result is False:
-            LOGGER.error(f'problem {present} Process<{process.pk}>')
-        elif isinstance(result, kiwipy.Future):
-            LOGGER.report(f'scheduled {infinitive} Process<{process.pk}>')
-            scheduled[result] = process
-        else:
-            LOGGER.error(f'got unexpected response when {present} Process<{process.pk}>: {result}')
+    if not timeout:
+        return
+
+    LOGGER.report(f"waiting for process(es) {','.join([str(proc.pk) for proc in futures.values()])}")
 
     try:
         for future, process in futures.items():
-            # unwrap is need here since LoopCommunicator will also wrap a future
+            # we unwrap to the end
             unwrapped = unwrap_kiwi_future(future)
             try:
-                result = unwrapped.result(timeout=timeout)
+                # future does not interpret float('inf') correctly by changing it to None we get the intended behavior
+                result = unwrapped.result(timeout=None if timeout == float('inf') else timeout)
             except communications.TimeoutError:
-                cancelled = unwrapped.cancel()
+                cancelled = future.cancel()
```

Comment:
I don't think this is correct here, the `future` is different from `unwrapped`.
Let's image the results are warped in 3 layers of future object, to get the result of final result from `future` requires `future.result().result().result()` but after unwrap to a flat future which is `unwrapped`, it requires only `unwrapped.result()`. It is the same for the `cancel()`.

Reply:
Nevermind, I think I am wrong here, the cancellation is taken care inside `unwrap_kiwi_future` where doing the cancellation when the inner most future is claim to be cancelled.